### PR TITLE
fail jobs if parameters are bigger than glidein site limit.

### LINKED ISSUE #5738: Do not submit jobs asking for too much resources
Now that glideInWMS is enforcing "MaxWallTimeMins" and "RequestDisk", we should change WMAgent NOT to submit greedy jobs to condor, otherwise those will be pending forever.
Instead, these jobs should be marked as jobfailed (or submitfailed) with a clear message.

It should be configurable and I guess it needs further discussion, but here is a first idea:
 - don't submit if job asks for more than 48h MaxWallTimeMins
 - don't submit if job asks for more than (27GB * RequestCpus) RequestDisk

### PR DESCRIPTION
**Pull request by ticoann**
first attempt to fix #5738. Alan, please review. I wanted to check this before job splitting but that seems hard. As Andrew suggested it has configurable parameter. Not sure about disk unit conversion. (Don't merge yet)

### CODE REVIEW

**Comment by ticoann** [comment]
@amaltaro, please review.

File: etc/WMAgentConfig.py [hunk 1]
```diff
@@ -181,6 +181,10 @@
 config.JobCreator.maxThreads = 1
 config.JobCreator.UpdateFromResourceControl = True
 config.JobCreator.pollInterval = 120
+config.JobCreator.CreateRestriction = {"GlindInCondition":
```
Comment (bbockelm) [inline etc/WMAgentConfig.py]: This is misspelled.

File: src/python/WMCore/JobSplitting/JobFactory.py [hunk 2]
```diff
@@ -420,3 +419,26 @@ def getPerformanceParameters(self, defaultParams):
         sizePerEvent = defaultParams.get('sizePerEvent', None) or 0
         memory = defaultParams.get('memoryRequirement', None) or 0
         return timePerEvent, sizePerEvent, memory
+    
+    def checkCreateFailConditoin(self, failedJob, failedReason):
```
Comment (bbockelm) [inline src/python/WMCore/JobSplitting/JobFactory.py]: This is misspelled.

File: src/python/WMCore/JobSplitting/JobFactory.py [hunk 3]
```diff
@@ -420,3 +419,26 @@ def getPerformanceParameters(self, defaultParams):
...
+
+        valueMap = {"MaxWallTimeHours": "estimatedJobTime", 
+                    "MaxMemory": "estimatedMemoryUsage",
```
Comment (bbockelm) [inline src/python/WMCore/JobSplitting/JobFactory.py]: Why not also include units on MaxMemory?

Comment (amaltaro) [inline src/python/WMCore/JobSplitting/JobFactory.py]: For the sake of clarity, I think we should use the same unit used in the classAds
 - MaxWallTimeMins (minutes)
 - RequestDisk (in kilobytes)

File: etc/WMAgentConfig.py (hunk 1 above)
Comment (amaltaro) [inline etc/WMAgentConfig.py]: can it be a single level dict?

File: src/python/WMCore/JobSplitting/JobFactory.py (hunk 2 above)
Comment (ticoann) [inline src/python/WMCore/JobSplitting/JobFactory.py]: Thanks Brian, I will fix that.

File: etc/WMAgentConfig.py (hunk 1 above)
Comment (ticoann) [inline etc/WMAgentConfig.py]: Thanks Guys I will change that.

File: src/python/WMCore/JobSplitting/JobFactory.py (hunk 3 above)
Comment (ticoann) [inline src/python/WMCore/JobSplitting/JobFactory.py]: Alan, I just change the unit to more human configurable. (I don't think they want to set by minutes)
But if you think it is more clear to use the same name and unit I will change that in the config. (and add comments which unit we are using)

It seems we are setting jdl key request_disk and request_memory instead of RequestDisk, Is it a bug?
Should it be RequestDisk and RequestMemory? Are both units KB?
https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/BossAir/Plugins/CondorPlugin.py#L1025

Comment (amaltaro) [inline src/python/WMCore/JobSplitting/JobFactory.py]: It's easier for us to read hours and GB, I agree! However, for eventual debugging and reading the code, it may be better to have the same unit used by condor. Anyways it's not a big deal, your call ;-)

About request_disk, I sent this question a long time ago to Tapas and I didn't get any answer back. To be honest, some magic happens and I don't know the trick :-D but I don't believe it's a bug since RequestDisk is nowhere in the code and the agent sets it properly.

Comment (bbockelm) [inline src/python/WMCore/JobSplitting/JobFactory.py]: In HTCondor submit files,

foo = 1

is a "submit file command".  It has certain macro expansion rules and can be manipulated into several attributes in the resulting classad.

+foo = 1

is a classad attribute - copied directly from the submit file (without the "+") to the ad.  No macro expansion is applied.

In the case of "request_disk", it is relatively straightforward converted to the attribute "RequestDisk".  The following two lines should be equivalent in a submit file:

request_disk = 1234
+RequestDisk = 1234

However, they are *not* the same in a classad (they are different attributes).

Comment (ericvaandering) [inline src/python/WMCore/JobSplitting/JobFactory.py]: Thanks Brian. I never knew the difference between these two forms. I'm still not sure I understand why and I am sure to forget next month, but still, thank you. :-)

Comment (amaltaro) [inline src/python/WMCore/JobSplitting/JobFactory.py]: I'm with you Eric, not sure I really got it :-)
Brian, if
> request_disk = 1234
> +RequestDisk = 1234 
are "equivalent" (but not the same in a classad (?)). Couldn't we replace this "request_disk" by "+RequestDisk" in the WMAgent code? Would you provide the same results in the end?

Comment (bbockelm) [inline src/python/WMCore/JobSplitting/JobFactory.py]: Yes, this would result in the same thing.

However, to illustrate the macro expansion,

foo = 1234
request_disk = $(foo)

is valid, while:

foo = 1234
+RequestDisk = $(foo)

is not.

Another difference is the handling of quoting rules.  The following two are *not* equivalent:

error = foo.err
+Err = foo.err

as "foo.err" should be a string and needs quoting:

+Err = "foo.err"

(Historical note: the submit file format predates ClassAds.  The "+" syntax was adopted to give the flexibility of using either format.  Most "normal users" probably find the submit file format preferable.)

**Comment by ticoann** [comment]
Thanks Alan, Brian for the review. I updated code as you suggested. (and leave request_disk in (Py)CondorPlugin as it is). Please review.  - (Don't merge until tested)

**Comment by hufnagel** [comment]
Can we please not merge this until we also tested this in the Tier0 ? We have a mix of singlecore and multicore and some very high disk jobs in there that are completely normal and should not be affected by this. The limits either need to take this into account or we need a way to disable this for the Tier0.

**Comment by hufnagel** [comment]
Btw, do these limits take into account number of cores ? I don't see it in the code... Case in point, we are planning to support 100GB repack jobs and that should just work.

**Comment by ticoann** [comment]
Hi Dirk, it doesn't count number of cores. I will add that. By settig the WMAgentConfig as following. 
config.JobCreator.GlideInRestriction  = {}
It should work as before. (not checking any restriction in wmagent side).
Anyway, I won't merge until Tier0 and WMAgent side is tested.
Thanks,

**Comment by AndrewLevin** [comment]
H Seangchan,

Can you explain what this pull request does?

Thanks.

Andrew

**Comment by ticoann** [comment]
It set the configurable upper limit and if WMAgent calculation is bigger than the limit it will set the upper limit so it can submit the job.

"MaxWallTimeMins": 48 * 60, # unit minutes to sync with GlideIn value
"RequestDisk": 27 * 1000, # unit KB, 27G

**Comment by ticoann** [comment]
retest this please

**Comment by amaltaro** [comment]
Actually this patch will fail jobs asking for more resources, no? So jobs won't even be created.

**Comment by ticoann** [comment]
Yes, Alan you are right. I sorry it is been a while. (removing previous comments) Thanks,

**Comment by AndrewLevin** [comment]
Hi Seangchan,

Can you also make a lower limit for the MaxWallTimeMins?

For relval I think we want to make all of the jobs run with a fixed MaxWallTimeMins, so we need both an upper limit and a lower limit.

Will this apply to all kinds of jobs (normal, merge, logcollect, cleanup)?

Andrew

**Comment by hufnagel** [comment]
This patch won't work ! At the absolute very least it'll have to take multicore into account. You cannot just have fixed limits for this.

I also question any automatic failures of jobs, there should be an easy way to completely turn this off (and not just being smart in the settings) as the Tier0 almost never wants to automatically fail jobs.
IMO, any kind of data processing this won't fly.

We have been busy with MC for too long. If you run a ReReco campaign, you cannot just fail jobs like this.

I especially question our ability to fail jobs on resource estimates which we know can be completely wrong.

As you might know, I just got a patch merged for the Tier0 that "solves" this problem in a completely different way. It detects too high limits and caps them to get the jobs to run. If the limits were right the jobs will get failed by Condor. Yes, this means it'll take a while til they make it through their retries and fail out.

Maybe as a compromise, how about having this configurable per spec. By default this is off, only certain types of workflows (like MC) enabled this automated job killing ?

**Comment by amaltaro** [comment]
If the WF team thinks it would be better to cap the resource limits instead of failing the jobs, then it's Ok with me as well.
At least for Central Production, usage estimates fall very close to reality.

Setting at spec level would give more flexibility, but in the end it may *just* not be used, and instead cause confusions from time to time.

**Comment by ericvaandering** [comment]
I'm inclined to side with @hufnagel on this one.

**Comment by amaltaro** [comment]
So, the suggestion is to change it to:
 - consider the number of cores
 - cap at close to 100% (maybe 46h and 25GB)?

**Comment by ericvaandering** [comment]
Why would you cap it at something lower? Those limits are already set a bit under the site limits.

**Comment by hufnagel** [comment]
Tier0 caps at 47h and 80GB or nCores*20GB (whatever is larger). There are generic code methods  for that that could be used elsewhere, see

https://github.com/dmwm/WMCore/pull/6077

Now, the Tier0 caps are to make sure we always can run a job at the T0 or any T1, they take the resources there into account (which are 8 core pilots with a minimum lifetime of 48h).

Generic limits might have to be stricter to ensure you can actually run the job.

Basically I set these limits because IMO the resource estimates are strictly to give glideInWMS some hints to better schedule jobs to active pilots. The resource estimates should never prevent a job from running at all.

If we want to kill jobs if they exceed certain values, we should do that from within glideInWMS (like we do now) when the job *actually* exceeds the limits, not based on an estimate that might or might not be correct.

**Comment by ericvaandering** [comment]
It seems to me this might be a good use for LogDB. If you are creating jobs that may not be able to complete in the allotted time, you may want a warning.




Eric W. Vaandering
Computing Physics Developer
CMS Computing Development Lead

Scientific Computing Division
Fermi National Accelerator Laboratory
630 840 6678 office
www.fnal.gov<http://www.fnal.gov/>

**Comment by bbockelm** [comment]
Uh, no, HTCondor won't start jobs (and GlideinWMS won't launch new pilots, although that's not an issue at the T0 probably) if the walltime or memory requested is too large.  The jobs will just hang out indefinitely until a timeout occurs or an operator removes them.

Since this is a per-agent configuration value, why does the Tier-0 particularly care?  Simply just increase the limit to something well-above plausible jobs.

Finally - CRAB3 approaches this by pulling in & caching the maximum configured walltime & memory from the relevant factory entries. There might be an analogous approach here.

**Comment by hufnagel** [comment]
Brian, the problem is that we cannot in general believe the limits we calculate. Especially for the Tier0 any event and size estimates for events, which maybe ok for the average, can be *HUGELY* off for some data taking conditions (depending on what the HLT is doing). That doesn't mean the data is bad, just different.

Therefore the job estimates for disk and wall time can be used to help scheduling, but they cannot be used to decide that the job can't possible complete. They aren't reliable enough for that. IMHO, this isn't only true for Tier0, but also for any other kind of real data processing (in the WMAgent). Data is not MC. That is the reason I can't have this for the Tier0. I could of course configure the Tier0 agent to effectively disable this, but I also think this won't fly for agent data processing.

Therefore the Tier0 follows the approach that any job we submit to HTCondor has to be runnable by the system, which means the disk and wall time limits we assign need to be lower than the pilot imposed limits for the resources we submit the job too. For the Tier0 this is relatively easy as we only access multicore pilots at the T0 and the T1. For general agent operations where jobs can run anywhere things are a bit more difficult.

As for the per-agent limit, even if we go this way and then maybe have special agents just for data processing where this configured differently, the max size limit needs to take singlecore/multicore into account. Multicore jobs will use more disk, you can't just configure a number.

But really, I would feel better if we could forget about this whole business of trying to kill jobs before even trying to run them because we *think* they'll fail based on some estimates someone made.

**Comment by AndrewLevin** [comment]
Hi,

I fully agree with Dirk. It is not realistic that we get even ballpark correct time-per-job estimates for release validation workflows also.

I would go even farther and suggest that glideinwms completely stop using wall time estimates. It seems to cause enormous headaches for everyone and I have not seen any evidence that there is any benefit from using it. 

At a minimum we need a way to make sure jobs don't get either stuck in pending or evicted by a dying pilot.

Andrew

**Comment by bbockelm** [comment]
Hi guys,

Maybe the mistake is thinking in the wrong terms here.  Don't provide a resource estimate to the system, provide a resource request.  If you go over the resource request (either in memory or walltime), then you incur the possibility the site will kill your job.

In our current setup, the resource requests to HTCondor result in a minimal guarantee from the system.  If you ask for a minimal guarantee runtime of 5 weeks, then likely the submit infrastructure won't be able to satisfy it.

One approach (which I believe Andrew has used) is to do something like:

resource request = min(upper bound, max(lower bound, resource estimate))

Brian

**Comment by amaltaro** [comment]
Just reminding everyone the final goal of this patch (or discussion) is NOT to have jobs pending forever.

Having said that, we could break this issue in 2 possible ways:
 1. let glideinWMS handle the resource request/estimate by:
  1.1. either nicely kill the job if nobody can serve it
  1.2. or send these jobs to the the best resources it can get (inside the desired_sites, of course) - one could also read qedit the classAds 
  1.3. as Andrew mentioned, completely remove this (which I'm not really in favor of)
 2. let WMAgent handle the resource request/estimate by:
  2.1. don't create jobs if they request too much resources (works well for Central Production and would even help to spot bugs or Ops mistakes)
  2.2. cap resource request based on the maximum available out there (which is not homogeneous)

In summary, we *must* either kill jobs or set them inside limits, each of them has its pros/cons, of course :-)
IF we go for setting boundaries, we could also make sure the "performance" error codes are set in the agents so they don't retry a) jobs running longer then pilot lifetime, b) taking too much disk resources c) taking too much RSS memory. I believe all of them are already set.

**Comment by hufnagel** [comment]
Just chatted with Alan about it. I think assuming working protections to prevent a job from being retried if it exhausts disk or wall time limits, we should at least try to run the jobs once, ie. err on the side of not believing the limits we calculate and make the job fit the resource constraints.

In practice this means we should never make a resource request that exceeds 47h or 46h of wall time and 20GB x numberOfCores of disk space. These would be the caps for the first submission of a job, what Brian calls the upper bound. I also would like to have the disk limits be configurable, because for the Tier0 we would like the option of requesting more than 20GB per core for a job (even if that means we can't fill all the cores in a multicore pilot).

The kill limits are somewhat independent of that, I think we still kill on 27GB of disk (20GB x numberOfCores for CERN AI). For the wall time what exactly is the policy ? Do we actively kill a job or do we let Condor kill it (actively or indirectly when the pilot dies). Do we prevent Condor rescheduling of such jobs ? I know we have various knobs in the system, would be good to list them all and make sure they make sense together.

As for what Brian called the lower bound, this would help if we assume that our resource estimates can be too low and want to avoid being scheduled on a pilot with a lifetime that is too short. Not sure how useful this is though. We can't set it too high because then we effectively can't run anything on short-lived pilots anymore and we also don't want to make this configurable by workflow (adds too much complexity). Maybe one hour would be good here (keep in mind we do have merge, cleanup, logcollect, etc jobs that can be very short) ? Alan speculated that Condor is ignoring too small wall time requests anyways and assumes a lower bound already, is this true ?

**Comment by AndrewLevin** [comment]
Also, can we set the wmagent performancekill timeout per job so that the job gets killed cleanly by wmagent when it is about to exceed its requested walltime?

**Comment by bbockelm** [comment]
HTCondor does not ignore the too-small wall time requests.

The potential danger in too-small requests is that you're only guaranteed the requested wall time.

Let's say the site batch system kills at 48 hours.  If the pilot has been around for 30 hours, you request 2 hours, and the job runs for 20 hours -- the batch system will kill the pilot (which tends to have the least amount of debugging of all failure modes).

**Comment by hufnagel** [comment]
If you request 2h and run for 20h, you take your chances. I don't think we want to have a lower bound of more than one or two hours, because then there would be no jobs to be able to run in an expiring pilot. For singlecore pilots this would just mean pilots die quicker, but for multicore pilots you would have serious efficiency problems.

**Comment by AndrewLevin** [comment]
Regarding the "least amount of debugging", we should make sure that wmruntime kills the job itself when it runs longer than its walltime request.

**Comment by hufnagel** [comment]
For the Tier0 this is a complete no-no. Even for the WMAgent I would say this should not be done, for instance any kind of ReReco is not going to work with this.

**Comment by AndrewLevin** [comment]
Sorry, why? As Brian said, the walltime estimate should be considered a resource request. When the job exceeds the resource request, the batch system or glideinwms is free to kill the job whenever it wants. All I am proposing is that we kill the job cleanly with wmruntime so we get back debugging information.

**Comment by amaltaro** [comment]
Andrew, you are talking about a different issue. Can you open another issue for that?
Here we just want to solve the resource estimation/request stuff.

**Comment by hufnagel** [comment]
No. For data processing the estimates aren't reliable enough. Yes, its a resource request, buts its a *minimal* resource request, not something we can kill on.

As I said before, for data processing we just can't predict job wall times good enough. I can make an estimate that *maybe* can help glideInWMS schedule the jobs better to pilots based on remaining pilot lifetimes, but at the end of the day it has to let the job run as long as it wants to (or the pilot allows).

**Comment by AndrewLevin** [comment]
As Brian said, right now glideinwms is not using the walltimes we provide as an estimate, but as a resource request. Therefore, we need to provide glideinwms with the time when it is free to kill the job.

**Comment by hufnagel** [comment]
No, we need two separate values for that. One to schedule on, another to kill on. They can't be the same. If they are, I am going to submit all Tier0 jobs with 47h requests.

**Comment by AndrewLevin** [comment]
That is what I am going to do for relval. (But probably I will set it to be 36 hours not 48.)

**Comment by hufnagel** [comment]
The problem with this approach is that you loose the benefit of trying to schedule the job to the right pilot and you also cannot do this for a workflow that is taking over a site (like PromptReco or ReReco can do). If all jobs pending for a site ask for 47h, how is that going to work ? At the very least it's leading to very inefficient use of resources. That's why I want to make the best estimate I can and only use it for scheduling, NOT for killing.

**Comment by AndrewLevin** [comment]
Yes, the walltime matching can lead to priority inversions, but I think Brian has been monitoring this and found that in practice it is rare. In any case, currently glideinwms does not support walltime estimates, only walltime resource requests. I don't see how can change this on the wmcore side. Maybe it would be more appropriate to make a github issue for glideinwms and discuss there.

**Comment by hufnagel** [comment]
Whether you call it estimate or request is semantics in the end. At the moment, AFAIK, glideInWMS does not call on the requested wall time, it only uses that information to schedule jobs. As long as that behavior stays like that I am fine with it. If they start killing on it, I would have to effectively completely disable that functionality for the Tier0 (even if that means I don't get any scheduling on wall time anymore).

**Comment by AndrewLevin** [comment]
scheduling on and killing on are the same thing, since if you schedule a 10 hour job to a pilot that has a  48 hour lifetime and has already been running for 40 hours, it will die after 8 hours and evict the job

**Comment by hufnagel** [comment]
They aren't the same thing. One is guaranteed, the other you take your chances. If you have a job with an 2h wall estimate and your pilots run for 48h, you are more than likely to get a pilot that has more than 2h left.

Especially for the Tier0 we used to have one month pilots. At the moment we are back to 48h but that will change again to one week.

**Comment by AndrewLevin** [comment]
I think we don't want to take our chances, especially for time-critical workflows like prompt reco and relval. 

If you have one month pilots, then you can get like 95% of the site by setting the walltime request to 48 hours. I think that is close enough to "taking over" for any practical purpose.

**Comment by hufnagel** [comment]
Speak for your own workflows :-). If the choice is between taking my chances and killing the job, I am taking my chances. All im saying here is the only way this feature is useful for me is if we do not kill on the requested wall time.

**Comment by hufnagel** [comment]
@ericvaandering , proposal is to calculate the job wall time estimate as normal, but then when setting the job request for wall time, to never ask for more than 47 hours and to always ask for a minimum of 1 hour. Technical implementation can be discussed, but if we go for something that applies to every job, we might as well make our life easy and implement the min and max in the Condor and PyCondorPlugin. Two lines of code there.

If we want to restrict this to certain jobs or workflows or have different behavior for different workflows, or if we want to be able to enable/disable this, we are looking at a different ball game.

IMO, we should never submit jobs that can never run, so this 1h min and 47h max should always be safe. For other numbers or for auto-killing of jobs that according to their estimate would run too long we need a configurable option though (as I already know that the Tier0 won't be able to use it).

On top of this there were discussion what to do with jobs that go over their estimate (kill them automatically, stop retries etc), but I think this should be a separate issue. For now and to solve the problem that triggered this issue, its sufficient if we change things so we don't get submit jobs to Condor that can never run.

**Comment by ericvaandering** [comment]
Right. This seems fine to me, but I don't see where the minimum is configured or set for a job. Can we do that too? Seems to me like an hour is asking for trouble if things get hung up, but as long as these are tunable in the agent config, we can do what we like.

How to handle a 8 hour estimated job that consistently goes over seems to me to be another issue, one that is not made better or worse by this patch.

**Comment by AndrewLevin** [comment]
Yes, I think we should have both a minimum and a maximum configured in the wmagent config.py

This issue can make that better or worse, since if the 8 hour estimate is replaced by 48 hours, then it will be better.

**Comment by hufnagel** [comment]
The minimum isn't even needed. If someone asks for a timePerEvent that is way too low, they'll take their chances being matched to a pilot that may or may not have time to finish the job. Tough luck. If they get matched to a "short" pilot and get killed, Condor will reschedule the job and try again on another pilot. I really don't see the problem.

The only way to avoid this is to pick a high number for the minimum, but if we do that for all jobs, how are we ever going to match anything to pilots that only have a few hours left ? With our infrastructure switching to multicore pilots this is not a way you want to run. Also the sites would prefer keeping pilots busy (even if that means some retries) instead of keeping them idle.

The 1h is basically a baseline, we don't expect we'll have any shorter jobs for production.

**Comment by hufnagel** [comment]
But sure, make it configurable for the CondorPlugin and PyCondorPlugin, that's usable. Tier0 and anyone running at CERN and the T1 should have a low minimum, for T2 it (currently) doesn't matter that much, higher minimums would also work.

**Comment by AndrewLevin** [comment]
No, the minimum is needed. For relval at least. I don't want jobs getting killed by condor. We need to report to requestors why the job got killed, and we can't just tell them "your job was killed by condor". Jobs can get killed by condor for many reasons.

**Comment by hufnagel** [comment]
RelVal should be contained to their special agent then, with different settings.

**Comment by AndrewLevin** [comment]
Yes, we have a relval team which has a dedicated agent.

**Comment by hufnagel** [comment]
@ticoann , the proposed code change here is completely different than what was proposed in the original subject now. You want to update the pull request or create a new one ?

**Comment by amaltaro** [comment]
So many replies here... I'll try to summarize it to make it clear:
 - make MaxWallTimeMins and RequestDisk configurable by agent (and taking into account # of cores)
 - if job goes beyond max resources, then set it to the configured limit ( * cores)
 - add the exit codes to the blacklist (don't retry jobs performance killed)
Bonus:
 - also add a MinMaxWallTimeMins, such that *any* job would have at least this estimation.

Is this summary correct?

**Comment by ericvaandering** [comment]
Correct, IMO

**Comment by hufnagel** [comment]
As long as adding the exit code is configurable too (which it is I think).

**Comment by hufnagel** [comment]
@amaltaro , which exit codes are you referring to here exactly ? Killed due to exceeding requested disk ? Thing is, at the moment the kill on disk is not done based on what the job requests AFAIK, these are different scripts.

**Comment by AndrewLevin** [comment]
Hi Alan,

I think we should not multiply by number cores. But just copy the max or min to the jdl. Dirk, is it ok with you?

Which exit code are you talking about? 61304? Isn't it already in the blacklist?

When a job is killed by condor it will not gave this exit code.

Andrew

**Comment by amaltaro** [comment]
Dirk, killing for exceeding time. I don't think there is a kill on disk usage anywhere but in the batch system.
Andrew, you are correct. Multiplication by # of cores for the requestDisk only, not for walltime.

I don't have the exact exit codes from the top of my head, but it's configurable.

> When a job is killed by condor it will not gave this exit code.

This is another issue and can be solved with proper assignment settings, at least for WallTime.
Anyways, I suggest to discuss it somewhere to leave a "clean" summary in the tail of this issue.

**Comment by ericvaandering** [comment]
retest this please

**Comment by amaltaro** [comment]
@ticoann I think this was taken care in #6201. Shall we close this one?

**Comment by amaltaro** [comment]
We decided in the past that we would not fail jobs going beyond the glidein limits. That was accomplished in this PR https://github.com/dmwm/WMCore/pull/6201

Shall we close this one?

**Comment by AndrewLevin** [comment]
yes

**Comment by amaltaro** [comment]
Closing, feel free to reopen it if you wish
